tooling/nargo/src/ops/compile.rs

```diff
@@ -0,0 +1,154 @@
+use acvm::{acir::circuit::Opcode, Language};
+use fm::FileManager;
+use noirc_driver::{CompilationResult, CompileOptions, CompiledContract, CompiledProgram};
+
+use crate::errors::CompileError;
+use crate::prepare_package;
+use crate::{package::Package, workspace::Workspace};
+
+use rayon::prelude::*;
+
+/// Compiles workspace.
+///
+/// # Errors
+///
+/// This function will return an error if .
+pub fn compile_workspace(
+    workspace: &Workspace,
+    binary_packages: &[Package],
+    contract_packages: &[Package],
+    np_language: Language,
+    is_opcode_supported: impl Fn(&Opcode) -> bool + std::marker::Sync,
+    compile_options: &CompileOptions,
+) -> Result<(Vec<CompiledProgram>, Vec<CompiledContract>), CompileError> {
+    // Compile all of the packages in parallel.
+    let program_results: Vec<(FileManager, CompilationResult<CompiledProgram>)> = binary_packages
+        .par_iter()
+        .map(|package| {
+            compile_program(workspace, package, compile_options, np_language, &is_opcode_supported)
+        })
+        .collect();
+    let contract_results: Vec<(FileManager, CompilationResult<CompiledContract>)> =
+        contract_packages
+            .par_iter()
+            .map(|package| {
+                compile_contract(package, compile_options, np_language, &is_opcode_supported)
+            })
+            .collect();
+
+    // Report any warnings/errors which were encountered during compilation.
+    let compiled_programs: Vec<CompiledProgram> = program_results
+        .into_iter()
+        .map(|(file_manager, compilation_result)| {
+            report_errors(
+                compilation_result,
+                &file_manager,
+                compile_options.deny_warnings,
+                compile_options.silence_warnings,
+            )
+        })
+        .collect::<Result<_, _>>()?;
+    let compiled_contracts: Vec<CompiledContract> = contract_results
+        .into_iter()
+        .map(|(file_manager, compilation_result)| {
+            report_errors(
+                compilation_result,
+                &file_manager,
+                compile_options.deny_warnings,
+                compile_options.silence_warnings,
+            )
+        })
+        .collect::<Result<_, _>>()?;
+
+    Ok((compiled_programs, compiled_contracts))
+}
+
+pub fn compile_program(
```

Comment:
is this code duplicated with cargo_cli?
if yes, can we have it only in one place?

Reply:
yes and no, cargo_cli also considers caching while this does not. It looks like we have more refactoring on the way and for that reason I decided to go with "as is" for time being to deliver something working and address debt soonish, would offer #3503 instead?